Add developer document describing system architecture

martyngigg opened the pull request:
**Description of work.**

Adds a refreshed and more detailed document dscribing the architecture of the codebase. It is a more modern and updated version of the original [ADD](https://github.com/mantidproject/documents/blob/a8e584151871bb61b4db2497af2e4cd5dc0f32da/Design/ArchitectureDesignDocument.doc).

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

Review the document and please make suggestions for improving it's clarity.

*There is no associated issue.*

*This does not require release notes* because **it is developer documentation**

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

martyngigg committed:
Add developer document describing system architecture

Co-authored-by: Gagik Vardanyan <vardanyan@ill.fr>
Co-authored-by: DavidFair <david.fairbrother@stfc.ac.uk>

shfoihoih

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,180 @@
+Summary
+-------
+
+Mantid is used for reduction and analysis of neutron and muon data.
```
I'm in favour of not forgetting the data visualization aspect, I think it's important to put it in the opening sentence.

martyngigg replied:
Good point. We spend a lot of time writing those!

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,180 @@
+
+   Context diagram for mantid software
+
+Experimental users of at neutron/muon facilities use the software
```
trailing `of`

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,180 @@
+
+The mantid software system is defined as four containers:
+two libraries (``framework`` & ``mantidqt``),
+one general-purpose graphical application (``workbench``) and
```
```suggestion
one general-purpose, cross-platform, graphical desktop application
```

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,180 @@
+cannot contain dependencies on those later, e.g.
+
+- framework *does not* depend on ``mantidqt`` or ``workbench``
+- ``mantidqt`` *does not* depend on ``workbench``
```
```suggestion
- ``mantidqt`` *does not* depend on ``workbench`` but does depend on ``framework``
```

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,180 @@
+
+   Container diagram for mantidqt Python package
+
+``mantidqt`` is a Python package defined house a collcetion of Qt widgets
```
there are typos here

martyngigg replied:
Thank you. Looking back it's clear I didn't read that part over again!

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,180 @@
+``workbench`` is an application written in pure Python in the ``qt/applications/workbench``
+directory of the source.
+It is not intended to be a library imported by others and as such offers
+so guarantees of API stability.
```
```suggestion
no guarantees of API stability.
```

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,180 @@
+such as the about screen, settings screen and project recovery mechanism that
+is specific to workbench.
+
+Plotting within ``workbench`` uses matplotlib_.
```
Shall we mention somewhere the OpenGL used in the instrument viewer? Or perhaps a small paragraph can mention that apart from plotting with matplotlib there are 2 advanced widgets: sliceviewer and instrument viewer.

martyngigg replied:
I might do this in the `mantidqt` and emphasize that it's not just basic thinks like text fields and combo boxes.

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,180 @@
+   :align: center
+   :alt: Container diagram of mantid software system
+
+   Container diagram for mantid software
```
It's unclear if the workbench depends on the custom interfaces. They are not really necessary for workbench, still they are built when building the workbench and appear in the menus.

martyngigg replied:
They are not strict dependencies from the code side, they are basically plugins as we do the registration and discover at run time. I will try to make this clearer in the diagram.

gvardany reviewed (commented):
Thanks for turning the ADD into a living document, this reflects how the project is organized currently.
Could you perhaps make the diagrams a little larger, maybe only the first two? The text was barely readable on those.

martyngigg reviewed (commented):
Thanks for the comments @gvardany and sorry for the delay in getting back to you with the comments.

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,186 @@
+
+The mantid software system is defined as four containers:
+two libraries (``framework`` & ``mantidqt``),
+one general-purpose, cross-platform, graphical desktop application
```
Is there a comma missing here?

martyngigg replied:
I'm missing an `and` I think.

gvardany commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,186 @@
+The aim of these customizations is to be transparent to a user in terms of the
+matplotlib api.
+A plotting script should function, without modification,
+outside of workbench with the standard matplotlib backends,
```
```suggestion
outside of workbench with the standard matplotlib backends, including jupyter notebooks
```
Perhaps we could also make explicit in the Framework/Python API section, that mantid as a framework can be used in vanilla python in batch mode or in any interactive python environment, including jupyter notebooks.

martyngigg replied:
Yes that's true. Being more explicit is worth it here.

gvardany commented on dev-docs/source/images/architecture-container.drawio:
```diff
@@ -0,0 +1 @@
+<mxfile host="Electron" modified="2021-07-19T14:00:48.510Z" agent="5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) draw.io/14.6.13 Chrome/89.0.4389.128 Electron/12.0.7 Safari/537.36" etag="ShSnSfSeEZ8RPLsCfOVm" version="14.6.13" type="device"><diagram id="prtHgNgQTEPvFCAcTncT" name="Page-1">5Vzbctq6Gn6azKx1QcZnw2XIaWcNbdKSNO2+2SOMABfbcmQ5QJ9+SbZ8lo0DhpBs2plgWZal//D9 ...
```
Github marks the io files not having a new line at the end, I don't think this is critical though.

martyngigg replied:
I'd prefer to keep those as saved by drawio as if it's updated then it'll just get overwritten.

gvardany reviewed (commented):
Thanks for the changes, just a couple of more remarks:

martyngigg committed:
Fix warning in dev-docs not including CodeCoverage

This page is not ready for the main contents page
but including it in the hidden TOC stops a warning
being produced.

martyngigg commented:
Thanks, Gagik. I've addressed those few issues.

DavidFair commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,194 @@
+Context
+-------
+
+The mantid software interacts with people and other systems as described
```
Nit this sentence is a bit clunky
```suggestion
People and other systems interact with the software package called Mantid as described
```

DavidFair commented on dev-docs/source/Architecture.rst:
```diff
@@ -0,0 +1,194 @@
+   Container diagram for workbench application
+
+``workbench`` is an application written in pure Python in the ``qt/applications/workbench``
+directory of the source.
```
Reading through we don't explicitly state that Workbench is our main "entrypoint" which allows users to do the majority of their work from a GUI perspective (that isn't an external GUI). I'm not sure if we want to state it here alongside the library section, or above where we introduce this?

```suggestion
directory of the source. It is intended to be the main interface provided by the project and allows access to the customized interfaces as described in `Container<container_>`.
```
Feel free to adapt this further as I'm not 100% how to phrase it

martyngigg replied:
Yeah this is a good point. Developers reading this should have an understanding of how the user community uses things.

DavidFair reviewed (changes requested):
Some minor nits on clarity. Otherwise builds correctly for me, renders correctly and reads well

martyngigg commented:
Thanks, David. I've addressed those couple of points.

DavidFair reviewed (approved):
Thanks for doing those nit picks :+1:
